content/blog/k8s-122-challenge/index.md

```diff
@@ -0,0 +1,51 @@
+
+## Another Challenge: Updating Active Deployments
+Although updating to StatefulSet NATS/etcd deployments is only necessary for Kubernetes 1.22+, we wanted to reduce our operational burden by updating all running Viziers.
+We encountered two issues:
+1. We did not have a path to update our dependencies (etcd/NATS).
+2. We had trouble including the recommended clients for the etcd and NATS custom resources.
+
+### Updating the dependencies
+Pixie’s Vizier is deployed and updated using the Kubernetes [operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/). The Vizier operator already had an update path for our core product, but lacked the equivalent for our dependencies. We originally delayed building the update path for the dependencies because we didn’t want to interrupt connected Viziers for longer than necessary and our deps rarely changed. This would also be the first update with destructive changes: we would need to get rid of the actively running etcd and NATS instances.
+
+We decided to special case this in the Vizier operator code. Whenever we detect the etcd-cluster and nats-cluster CustomResources in the Vizier namespace, we remove the resources and deploy the new StatefulSet versions. We didn’t create an update path for dependencies, but given the destructive changes were a special case, we decided to delay building out a whole update path.
+
+### Including the etcd and NATS clients
+We wanted to use the etcd-cluster and nats-cluster clients to detect old NATS and etcd dependencies. We hit a [strange compilation error](https://github.com/coreos/etcd-operator/issues/2167) while attempting to include the clients. It turns out that our packaged version of client-go is incompatible with the packaged version in the clients. The client-go developers introduced a [breaking interface change](https://github.com/kubernetes/client-go/commit/ae9f6b2601c8f8b97ad2865943f423d65539ffdb) and Go’s dependency management disallows [two different incompatible versions](https://research.swtch.com/vgo-import#dependency_story) with the same import path.
+
+Fortunately, we had recently used `client-gen` [to generate client code for Vizier CustomResources](https://github.com/pixie-io/pixie/commit/3950a03d723029b4280e5267f0c2c9c78f2dd7c8). We decided to do the same with nats-cluster and etcd-cluster, [vendoring the clients into our operator code](https://github.com/pixie-io/pixie/search?q=generate&type=commits). Once you have a hammer, everything starts to look like a nail.
```

Comment:
vendoring link should be https://github.com/pixie-io/pixie/commit/c5079278c55c70451a59d9fc315b56d311aa2e54

Reply:
like this exact format: `[pixie-io/pixie@c507927](https://github.com/pixie-io/pixie/commit/c5079278c55c70451a59d9fc315b56d311aa2e54)`